Reconstify `Add`

fee1-dead opened the pull request:
r? project-const-traits

I'm not happy with the ui test changes (or failures because I did not bless them and include the diffs in this PR). There is at least some bugs I need to look and try fix:

1. A third duplicated diagnostic when a consumer crate that does not have `effects` enabled has a trait selection error for an upstream const_trait trait. See tests/ui/ufcs/ufcs-qpath-self-mismatch.rs.
2. For some reason, making `Add` a const trait would stop us from suggesting `T: Add` when we try to add two `T`s without that bound. See tests/ui/suggestions/issue-97677.rs

fee1-dead committed:
Reconstify `Add`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7670070436/job/20905396251) [(plain)](https://github.com/rust-lang/rust/commit/8284875de841469120a719b11229a09968684182/checks/20905396251/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Fri Jan 26 15:33:14 UTC 2024
  network time: Fri, 26 Jan 2024 15:33:14 GMT
  network time: Fri, 26 Jan 2024 15:33:14 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
72    |              B
-    |
- help: consider restricting type parameter `B`
-    |
- LL | impl<B: std::ops::Add<Output = B>> Add for D<B> {
78 
79 error[E0308]: mismatched types
80   --> $DIR/missing-bounds.rs:42:14



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/missing-bounds/missing-bounds.stderr
diff of fixed:

24 
25 struct D<B>(B);
26 
- impl<B: std::ops::Add<Output = B>> Add for D<B> {
+ impl<B> Add for D<B> {
29 
30     fn add(self, rhs: Self) -> Self {


---
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/generic-associated-types/missing-bounds.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/missing-bounds" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generic-associated-types/missing-bounds/auxiliary"
stdout: none
--- stderr -------------------------------
error: equality constraints are not yet supported in `where` clauses
   |
   |
LL | impl<B: Add> Add for E<B> where <B as Add>::Output = B {
   |
   = note: see issue #20041 <https://github.com/rust-lang/rust/issues/20041> for more information
help: if `Output` is an associated type you're trying to set, use the associated type binding syntax
   |
   |
LL | impl<B: Add> Add for E<B> where B: Add<Output = B> {

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:11:11
   |
   |
LL | impl<B> Add for A<B> where B: Add {
   |      - expected this type parameter
...
LL |         A(self.0 + rhs.0) //~ ERROR mismatched types
   |         - ^^^^^^^^^^^^^^ expected type parameter `B`, found associated type
   |         arguments to this struct are incorrect
   |
   = note: expected type parameter `B`
   = note: expected type parameter `B`
             found associated type `<B as Add>::Output`
help: the type constructed contains `<B as Add>::Output` due to the type of the argument passed
   |
   |
LL |         A(self.0 + rhs.0) //~ ERROR mismatched types
   |         ^^--------------^
   |           this argument influences the type of `A`
note: tuple struct defined here
  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:5:8
   |
   |
LL | struct A<B>(B);
help: consider further restricting this bound
   |
   |
LL | impl<B> Add for A<B> where B: Add<Output = B> {

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:21:14
   |
   |
LL | impl<B: Add> Add for C<B> {
   |      - expected this type parameter
...
LL |         Self(self.0 + rhs.0) //~ ERROR mismatched types
   |         ---- ^^^^^^^^^^^^^^ expected type parameter `B`, found associated type
   |         arguments to this function are incorrect
   |
   = note: expected type parameter `B`
   = note: expected type parameter `B`
             found associated type `<B as Add>::Output`
Build completed unsuccessfully in 0:12:55
  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:15:8
   |
   |
LL | struct C<B>(B);
help: consider further restricting this bound
   |
   |
LL | impl<B: Add<Output = B>> Add for C<B> {


error[E0369]: cannot add `B` to `B`
   |
   |
LL |         Self(self.0 + rhs.0) //~ ERROR cannot add `B` to `B`
   |              ------ ^ ----- B
   |              B

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:42:14
##[error]  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:42:14
   |
LL | impl<B: Add> Add for E<B> where <B as Add>::Output = B {
   |      - expected this type parameter
...
LL |         Self(self.0 + rhs.0) //~ ERROR mismatched types
   |         ---- ^^^^^^^^^^^^^^ expected type parameter `B`, found associated type
   |         arguments to this function are incorrect
   |
   = note: expected type parameter `B`
   = note: expected type parameter `B`
             found associated type `<B as Add>::Output`
  --> /checkout/tests/ui/generic-associated-types/missing-bounds.rs:35:8
   |
   |
LL | struct E<B>(B);
help: consider further restricting this bound
   |
   |
LL | impl<B: Add<Output = B>> Add for E<B> where <B as Add>::Output = B {

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0308, E0369.
Some errors have detailed explanations: E0308, E0369.
For more information about an error, try `rustc --explain E0308`.
------------------------------------------


---- [ui] tests/ui/suggestions/issue-97677.rs stdout ----
diff of stderr:

5    |     - ^ -- {integer}
7    |     N
-    |
- help: consider restricting type parameter `N`
-    |
-    |
- LL | fn add_ten<N: std::ops::Add<i32, Output = N>>(n: N) -> N {
13 
14 error: aborting due to 1 previous error
15 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-97677/issue-97677.stderr
diff of fixed:

1 // run-rustfix
2 
- fn add_ten<N: std::ops::Add<i32, Output = N>>(n: N) -> N {
+ fn add_ten<N>(n: N) -> N {
4     n + 10
5     //~^ ERROR cannot add `{integer}` to `N`


The actual fixed differed from the expected fixed.
The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-97677/issue-97677.fixed
To only update this specific test, also pass `--test-args suggestions/issue-97677.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/suggestions/issue-97677.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-97677" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/issue-97677/auxiliary"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0369]: cannot add `{integer}` to `N`
   |
LL |     n + 10
LL |     n + 10
   |     - ^ -- {integer}
   |     N

error: aborting due to 1 previous error


For more information about this error, try `rustc --explain E0369`.
------------------------------------------


---- [ui] tests/ui/suggestions/restrict-type-not-param.rs stdout ----
diff of stderr:

13    | ^^^^^^^^^^^^^^^^^ must implement `Add<T>`
14 note: the trait `Add` must be implemented
15   --> $SRC_DIR/core/src/ops/arith.rs:LL:COL
- help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
-    |
- LL | fn qux<T>(a: Wrapper<T>, b: T) -> T where Wrapper<T>: Add<T, Output = T> {
20 
21 error: aborting due to 1 previous error
22 

---
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/suggestions/restrict-type-not-param.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/restrict-type-not-param" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/restrict-type-not-param/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0369]: cannot add `T` to `Wrapper<T>`
   |
LL |     a + b
LL |     a + b
   |     - ^ - T
   |     Wrapper<T>
   |
   |
note: an implementation of `Add<T>` might be missing for `Wrapper<T>`
   |
   |
LL | struct Wrapper<T>(T);
   | ^^^^^^^^^^^^^^^^^ must implement `Add<T>`
note: the trait `Add` must be implemented
  --> /rustc/FAKE_PREFIX/library/core/src/ops/arith.rs:77:1
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0369`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/type/type-check/missing_trait_impl.rs stdout ----
diff of stderr:

5    |             - ^ - T
7    |             T
-    |
- help: consider restricting type parameter `T`
-    |
-    |
- LL | fn foo<T: std::ops::Add>(x: T, y: T) {
13 
13 
14 error[E0368]: binary assignment operation `+=` cannot be applied to type `T`


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/type-check/missing_trait_impl/missing_trait_impl.stderr
---
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/type/type-check/missing_trait_impl.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/type-check/missing_trait_impl" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/type-check/missing_trait_impl/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0369]: cannot add `T` to `T`
   |
   |
LL |     let z = x + y; //~ ERROR cannot add `T` to `T`
   |             - ^ - T
   |             T


error[E0368]: binary assignment operation `+=` cannot be applied to type `T`
   |
   |
LL |     x += x; //~ ERROR binary assignment operation `+=` cannot be applied to type `T`
   |     -^^^^^
   |     |
   |     cannot use `+=` on type `T`
help: consider restricting type parameter `T`
   |
   |
LL | fn bar<T: std::ops::AddAssign>(x: T) {

error[E0600]: cannot apply unary operator `-` to type `T`
##[error]  --> /checkout/tests/ui/type/type-check/missing_trait_impl.rs:13:13
   |
   |
LL |     let y = -x; //~ ERROR cannot apply unary operator `-` to type `T`
   |             ^^ cannot apply unary operator `-`
help: consider restricting type parameter `T`
   |
   |
LL | fn baz<T: std::ops::Neg>(x: T) {

error[E0600]: cannot apply unary operator `!` to type `T`
##[error]  --> /checkout/tests/ui/type/type-check/missing_trait_impl.rs:14:13
   |
   |
LL |     let y = !x; //~ ERROR cannot apply unary operator `!` to type `T`
   |             ^^ cannot apply unary operator `!`
help: consider restricting type parameter `T`
   |
   |
LL | fn baz<T: std::ops::Not>(x: T) {

error[E0614]: type `T` cannot be dereferenced
##[error]  --> /checkout/tests/ui/type/type-check/missing_trait_impl.rs:15:13
   |
   |
LL |     let y = *x; //~ ERROR type `T` cannot be dereferenced

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0368, E0369, E0600, E0614.
Some errors have detailed explanations: E0368, E0369, E0600, E0614.
For more information about an error, try `rustc --explain E0368`.
------------------------------------------


---- [ui] tests/ui/ufcs/ufcs-qpath-self-mismatch.rs stdout ----
diff of stderr:

11              <&'a i32 as Add<i32>>
12              <&i32 as Add<&i32>>
13 
+ error[E0277]: cannot add `u32` to `i32`
+   --> $DIR/ufcs-qpath-self-mismatch.rs:4:5
+    |
+ LL |     <i32 as Add<u32>>::add(1, 2);
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `i32 + u32`
+    = help: the trait `Add<u32>` is not implemented for `i32`
+    = help: the trait `Add<u32>` is not implemented for `i32`
+    = help: the following other types implement trait `Add<Rhs>`:
+              <i32 as Add>
+              <i32 as Add<&i32>>
+              <&'a i32 as Add<i32>>
+              <&i32 as Add<&i32>>
14 error[E0308]: mismatched types
15   --> $DIR/ufcs-qpath-self-mismatch.rs:7:28
16    |


68              <&'a i32 as Add<i32>>
69              <&i32 as Add<&i32>>
- error: aborting due to 4 previous errors
+ error: aborting due to 5 previous errors
72 
73 Some errors have detailed explanations: E0277, E0308.
---
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/ufcs/ufcs-qpath-self-mismatch" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/ufcs/ufcs-qpath-self-mismatch/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0277]: cannot add `u32` to `i32`
   |
   |
LL |     <i32 as Add<u32>>::add(1, 2);
   |      ^^^ no implementation for `i32 + u32`
   = help: the trait `Add<u32>` is not implemented for `i32`
   = help: the trait `Add<u32>` is not implemented for `i32`
   = help: the following other types implement trait `Add<Rhs>`:
             <i32 as Add>
             <i32 as Add<&i32>>
             <&'a i32 as Add<i32>>
             <&i32 as Add<&i32>>

error[E0277]: cannot add `u32` to `i32`
   |
   |
LL |     <i32 as Add<u32>>::add(1, 2);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `i32 + u32`
   = help: the trait `Add<u32>` is not implemented for `i32`
   = help: the trait `Add<u32>` is not implemented for `i32`
   = help: the following other types implement trait `Add<Rhs>`:
             <i32 as Add>
             <i32 as Add<&i32>>
             <&'a i32 as Add<i32>>
             <&i32 as Add<&i32>>
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs:7:28
   |
   |
LL |     <i32 as Add<i32>>::add(1u32, 2);
   |     ---------------------- ^^^^ expected `i32`, found `u32`
   |     arguments to this function are incorrect
   |
help: the return type of this call is `u32` due to the type of the argument passed
  --> /checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs:7:5
  --> /checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs:7:5
   |
LL |     <i32 as Add<i32>>::add(1u32, 2);
   |                            |
   |                            this argument influences the return type of `Add`
note: method defined here
  --> /rustc/FAKE_PREFIX/library/core/src/ops/arith.rs:92:8
  --> /rustc/FAKE_PREFIX/library/core/src/ops/arith.rs:92:8
help: change the type of the numeric literal from `u32` to `i32`
   |
LL |     <i32 as Add<i32>>::add(1i32, 2);

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs:9:31
   |
   |
LL |     <i32 as Add<i32>>::add(1, 2u32);
   |     ----------------------    ^^^^ expected `i32`, found `u32`
   |     arguments to this function are incorrect
   |
help: the return type of this call is `u32` due to the type of the argument passed
  --> /checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs:9:5
  --> /checkout/tests/ui/ufcs/ufcs-qpath-self-mismatch.rs:9:5
   |
LL |     <i32 as Add<i32>>::add(1, 2u32);
   |                               |
   |                               this argument influences the return type of `Add`
note: method defined here
  --> /rustc/FAKE_PREFIX/library/core/src/ops/arith.rs:92:8
  --> /rustc/FAKE_PREFIX/library/core/src/ops/arith.rs:92:8
help: change the type of the numeric literal from `u32` to `i32`
   |
LL |     <i32 as Add<i32>>::add(1, 2i32);


error[E0277]: cannot add `u32` to `i32`
   |
   |
LL |     <i32 as Add<u32>>::add(1, 2);
   |     ^^^^^^^^^^^^^^^^^^^^^^ no implementation for `i32 + u32`
   = help: the trait `Add<u32>` is not implemented for `i32`
   = help: the trait `Add<u32>` is not implemented for `i32`
   = help: the following other types implement trait `Add<Rhs>`:
             <i32 as Add>
             <i32 as Add<&i32>>
             <&'a i32 as Add<i32>>
             <&i32 as Add<&i32>>
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
```

</details>

fee1-dead commented:
the first item seems to be reverting a previous const_trait_impl change: https://github.com/rust-lang/rust/pull/100982/commits/191d3b76db5ee0f7a05ff79b6d584dde5f6e404f#diff-bcbca6666787ee0e6f033b0ba775250deb6d0cfb58d2f2195e11635f2e00ad11

fee1-dead committed:
make effect infer variables suggestable in diagnostics

it works when a non-const context that does not enable effects
calls into a const effects-enabled trait. We'd simply suggest the
non-const trait bound in this case consistent to its fallback.

fee1-dead committed:
fix ui tests

these ui changes are closer to what was there before const_trait_impl changes.

fee1-dead commented:
making as ready for review, see the latest commits

fee1-dead commented:
.. forgot to rename the title

compiler-errors commented on tests/ui/ufcs/ufcs-qpath-self-mismatch.stderr:
```diff
@@ -11,16 +11,29 @@ LL |     <i32 as Add<u32>>::add(1, 2);
              <&'a i32 as Add<i32>>
              <&i32 as Add<&i32>>
 
+error[E0277]: cannot add `u32` to `i32`
```
It's fine that this error message is duplicated imo.

compiler-errors commented on tests/ui/suggestions/invalid-bin-op.stderr:
```diff
@@ -16,6 +16,10 @@ help: consider annotating `S<T>` with `#[derive(PartialEq)]`
    |
+help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
+   |
+LL | pub fn foo<T>(s: S<T>, t: S<T>) where S<T>: PartialEq {
```
Why is this being suggested now?

fee1-dead replied:
It was removed in a const-traits PR: https://github.com/rust-lang/rust/commit/d464dd07ed83cc25ee3e568c98e3d4e26ff364ae#diff-a4f9511c26556adec6fac6f4073f446a12e9c9c883b0fd5ae672956caa5836c9 in #118661. This has been suggested before `PartialEq` gained the effect param, and adding the effect param in that commit made `PartialEq` not suggestable (which this PR makes it suggestable again)

compiler-errors commented:
r=me on the changes, though I am curious why that new suggestion is being suggested. Can we suppress it?

fee1-dead commented:
I don't know why that suggestion exists, but I don't think it is related to const traits (per my latest comment).

@rustbot ready

compiler-errors commented:
cool

@bors r+

bors commented:
:pushpin: Commit 2e1e574ce6a4ed302330a653fa15d149857f0942 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 2e1e574ce6a4ed302330a653fa15d149857f0942 -->
<!-- homu: {"type":"Approved","sha":"2e1e574ce6a4ed302330a653fa15d149857f0942","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/120447#issuecomment-1920036755) for pull requests below priority 100. This pull request will be tested once the tree is reopened.

matthiaskrgr commented:
@bors r-
seems to be failing in https://github.com/rust-lang/rust/pull/120622#issuecomment-1925453384 ? 

can this have some perf impact, I wonder if we should rollup iffy or never.

fee1-dead commented:
I have blessed the test. Don't think this could have a large perf impact, but it might because of adding a generic param though.

@bors rollup=iffy

fee1-dead commented:
Blessing is a minor change, don't think it needs to be reviewed again

@bors r=compiler-errors

bors commented:
:pushpin: Commit 74dbf3a070adac13f28f39e2d447b0148998cc40 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 74dbf3a070adac13f28f39e2d447b0148998cc40 -->
<!-- homu: {"type":"Approved","sha":"74dbf3a070adac13f28f39e2d447b0148998cc40","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 74dbf3a070adac13f28f39e2d447b0148998cc40 with merge 6894f435d35d3d540dcefbc51390158ca5954861...
<!-- homu: {"type":"BuildStarted","head_sha":"74dbf3a070adac13f28f39e2d447b0148998cc40","merge_sha":"6894f435d35d3d540dcefbc51390158ca5954861"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7822788778/job/21345567556)
Approved by: compiler-errors
Pushing 6894f435d35d3d540dcefbc51390158ca5954861 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7822788778/job/21345567556"},"merge_sha":"6894f435d35d3d540dcefbc51390158ca5954861"} -->

rust-timer commented:
Finished benchmarking commit (6894f435d35d3d540dcefbc51390158ca5954861): [comparison URL](https://perf.rust-lang.org/compare.html?start=8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6&end=6894f435d35d3d540dcefbc51390158ca5954861&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6&end=6894f435d35d3d540dcefbc51390158ca5954861&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.4%, 0.5%] | 2     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.4%, 0.5%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6&end=6894f435d35d3d540dcefbc51390158ca5954861&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.1% | [-2.1%, -2.1%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6&end=6894f435d35d3d540dcefbc51390158ca5954861&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -8.7% | [-8.7%, -8.7%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6&end=6894f435d35d3d540dcefbc51390158ca5954861&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 662.945s -> 663.461s (0.08%)
**Artifact size**: 308.24 MiB -> 308.30 MiB (0.02%)
